File: src/engine/gapDetector.js
import { storage } from '../services/storage'
import categoryVocab from '../data/categoryVocab.json'

const GAP_SIGNALS_KEY = 'gap_signals_'
const GAP_ALERTS_KEY = 'gap_alerts_'
const SIGNAL_THRESHOLD = 3
const WINDOW_DAYS = 7
const BROWSE_TIMEOUT_MS = 8000

let browseTimer = null
let currentBoardId = null
let currentProfileId = null
let didTapOnBoard = false

export function startBrowseTimer(profileId, boardId) {
  clearBrowseTimer()
  currentBoardId = boardId
  currentProfileId = profileId
  didTapOnBoard = false

  browseTimer = setTimeout(() => {
    // Timer fired without a tap — record gap signal
    if (!didTapOnBoard) {
      recordGapSignal(profileId, boardId)
    }
  }, BROWSE_TIMEOUT_MS)
}

export function clearBrowseTimer() {
  if (browseTimer) {
    clearTimeout(browseTimer)
    browseTimer = null
  }
}

export function markTappedOnBoard() {
  didTapOnBoard = true
  clearBrowseTimer()
}

function recordGapSignal(profileId, boardId) {
  const key = `${GAP_SIGNALS_KEY}${profileId}`
  const signals = storage.get(key) || []
  signals.push({
    boardId,
    profileId,
    timestamp: new Date().toISOString()
  })
  storage.set(key, signals)
  evaluateGaps(profileId)
}

function evaluateGaps(profileId) {
  const key = `${GAP_SIGNALS_KEY}${profileId}`
  const signals = storage.get(key) || []

  const cutoff = new Date()
  cutoff.setDate(cutoff.getDate() - WINDOW_DAYS)
  const recent = signals.filter(s => new Date(s.timestamp) > cutoff)

  // Count signals per board
  const counts = {}
  recent.forEach(s => {
    counts[s.boardId] = (counts[s.boardId] || 0) + 1
  })

  // Generate alerts for boards over threshold
  const alertsKey = `${GAP_ALERTS_KEY}${profileId}`
  const existingAlerts = storage.get(alertsKey) || []

  Object.entries(counts).forEach(([boardId, count]) => {
    if (count >= SIGNAL_THRESHOLD) {
      const alreadyAlerted = existingAlerts.find(
        a => a.boardId === boardId && a.weekOf === getCurrentWeek()
      )
      if (!alreadyAlerted) {
        existingAlerts.push({
          boardId,
          profileId,
          signalCount: count,
          weekOf: getCurrentWeek(),
          createdAt: new Date().toISOString()
        })
      }
    }
  })

  storage.set(alertsKey, existingAlerts)
}

export function getGapAlerts(profileId, boards) {
  const alertsKey = `${GAP_ALERTS_KEY}${profileId}`
  const alerts = storage.get(alertsKey) || []
  const thisWeek = getCurrentWeek()

  return alerts
    .filter(a => a.weekOf === thisWeek)
    .map(alert => {
      const board = boards.find(b => b.id === alert.boardId)
      if (!board) return null
      const existingLabels = board.symbols.map(s => s.label.toLowerCase())
      const category = board.category || 'custom'
      const expansionList = categoryVocab[category] || categoryVocab.custom
      const suggestions = expansionList
        .filter(word => !existingLabels.includes(word.toLowerCase()))
        .slice(0, 5)
      return {
        ...alert,
        boardName: board.name,
        existingSymbolCount: board.symbols.length,
        suggestedAdditions: suggestions
      }
    })
    .filter(Boolean)
}

function getCurrentWeek() {
  const now = new Date()
  const start = new Date(now.getFullYear(), 0, 1)
  const week = Math.ceil(((now - start) / 86400000 + start.getDay() + 1) / 7)
  return `${now.getFullYear()}-W${week}`
}

File: backend/app/engines/gap_detector.py
import json
from datetime import datetime, timedelta, timezone
from pathlib import Path

from sqlalchemy import select
from sqlalchemy.orm import Session

from ..models import Board, GapAlert, GapSignal

SIGNAL_THRESHOLD = 3
WINDOW_DAYS = 7

_vocab_path = Path(__file__).resolve().parent.parent / "data" / "category_vocab.json"
CATEGORY_VOCAB: dict[str, list[str]] = json.loads(_vocab_path.read_text(encoding="utf-8"))


def current_week() -> str:
    now = datetime.now(timezone.utc)
    start = datetime(now.year, 1, 1, tzinfo=timezone.utc)
    week = int(((now - start).days + start.weekday() + 1) / 7) + 1
    return f"{now.year}-W{week}"


def record_gap_signal(db: Session, profile_id: str, board_id: str) -> None:
    db.add(GapSignal(profile_id=profile_id, board_id=board_id))
    db.flush()
    evaluate_gaps(db, profile_id)


def evaluate_gaps(db: Session, profile_id: str) -> None:
    cutoff = datetime.now(timezone.utc) - timedelta(days=WINDOW_DAYS)
    signals = db.scalars(
        select(GapSignal)
        .where(GapSignal.profile_id == profile_id)
        .where(GapSignal.timestamp > cutoff)
    ).all()

    counts: dict[str, int] = {}
    for s in signals:
        counts[s.board_id] = counts.get(s.board_id, 0) + 1

    week = current_week()
    for board_id, count in counts.items():
        if count < SIGNAL_THRESHOLD:
            continue
        existing = db.scalar(
            select(GapAlert)
            .where(GapAlert.profile_id == profile_id)
            .where(GapAlert.board_id == board_id)
            .where(GapAlert.week_of == week)
        )
        if existing:
            existing.signal_count = count
        else:
            db.add(GapAlert(
                profile_id=profile_id,
                board_id=board_id,
                signal_count=count,
                week_of=week,
            ))
    db.flush()


def get_gap_alerts(db: Session, profile_id: str) -> list[dict]:
    week = current_week()
    alerts = db.scalars(
        select(GapAlert)
        .where(GapAlert.profile_id == profile_id)
        .where(GapAlert.week_of == week)
    ).all()

    results = []
    for alert in alerts:
        board = db.get(Board, alert.board_id)
        if board is None:
            continue
        existing_labels = {s.label.lower() for s in board.symbols}
        expansion = CATEGORY_VOCAB.get(board.category, CATEGORY_VOCAB["custom"])
        suggestions = [w for w in expansion if w.lower() not in existing_labels][:5]
        results.append({
            "board_id": board.id,
            "board_name": board.name,
            "signal_count": alert.signal_count,
            "week_of": alert.week_of,
            "existing_symbol_count": len(board.symbols),
            "suggested_additions": suggestions,
        })
    return results
